haskell/src/Parser.hs

```diff
@@ -0,0 +1,43 @@
+module Parser (evaluateExpression) where
+
+import Text.ParserCombinators.Parsec
+import Nodes
+
+evaluateExpression  :: String -> Maybe ([Node])
+evaluateExpression input =
+    case parse parser "(unknown)" input of
+    Left parseError -> Nothing
+    Right result -> Just result
+
+parser :: Parser [Node]
+parser = expressionParser
+
+expressionParser :: Parser [Node]
+expressionParser = do
+    number <- numberParser
+    spaces
+    nodes <- many furtherExpressionParser
+    return $ [NumberNode (read number)] ++ (concat nodes)
+
+furtherExpressionParser :: Parser [Node]
+furtherExpressionParser = do
+    spaces
+    opChar <- operatorParser
+    spaces
+    number <- numberParser
+    return $ [OperatorNode (operatorFromChar opChar), NumberNode (read number)]
+
+integerParser :: Parser String
+integerParser = many digit
+
+numberParser :: Parser String
+numberParser = integerParser
+
+operatorParser :: Parser Char
```

Comment:
You can combine `operatorParser` and `operatorFromChar` in a way that doesn't require using `error` using `<|>`. I think it's something like:

``` haskell
operatorParser :: Parser Char
operatorParser = try (char '+' >> return Plus)
    <|> try (char '-' >> return Minus)
    <|> (char '*' >> return Times)
```

This will either map one of the matched characters to an enum value or result in a parser error.